src/xclient_zmq_impl.hpp

```diff
@@ -0,0 +1,105 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay, Martin Renou          *
+* Copyright (c) 2016, QuantStack                                           *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XEUS_CLIENT_ZMQ_IMPL_HPP
+#define XEUS_CLIENT_ZMQ_IMPL_HPP
+
+#include <nlohmann/json.hpp>
+#include "zmq.hpp"
+
+#include "xeus/xeus_context.hpp"
+#include "xeus/xkernel_configuration.hpp"
+#include "xeus/xmessage.hpp"
+
+#include "xeus-zmq/xthread.hpp"
+
+#include "xdealer_channel.hpp"
+#include "xiopub_client.hpp"
+#include "xclient_messenger.hpp"
+
+namespace xeus
+{
+    class xauthentication;
+
+    class xclient_zmq_impl
+    {
+    public:
+        using iopub_client_ptr = std::unique_ptr<xiopub_client>;
+        using listener = std::function<void(xmessage)>;
+
+        xclient_zmq_impl(zmq::context_t& context,
+                    const xconfiguration& config,
+                    nl::json::error_handler_t eh);
+
+        ~xclient_zmq_impl();
+
+        xclient_zmq_impl(const xclient_zmq_impl&) = delete;
+        xclient_zmq_impl& operator=(const xclient_zmq_impl&) = delete;
+
+        xclient_zmq_impl(xclient_zmq_impl&&) = delete;
+        xclient_zmq_impl& operator=(xclient_zmq_impl&&) = delete;
+
+        // shell channel
+        void send_on_shell(xmessage msg);
+        std::optional<xmessage> receive_on_shell(long timeout);
+        void register_shell_listener(const listener& l);
+
+        // control channel
+        void send_on_control(xmessage msg);
+        std::optional<xmessage> receive_on_control(long timeout);
+        void register_control_listener(const listener& l);
+
+        // iopub channel
+        std::size_t iopub_queue_size() const;
+        std::optional<xmessage> pop_iopub_message();
+        void register_iopub_listener(const listener& l);
+
+        // hearbeat channel
+        // TODO
+
+        // client messenger
+        xclient_messenger& get_client_messenger();
```

Comment:
Not sure if we would be using this anywhere.

Reply:
I think this should not be exposed, this is for internal synchronization between the different threads. It might be useful as a protected method if we implement other client models in the future, but for now it's safe to not expose it.

Reply:
Noted.